CaloCluster/src/CaloClusterFast_module.cc

```diff
@@ -1,322 +1,197 @@
-           const auto& waveform = caloDigi.waveform();
-
-           auto it   = waveform.begin();
-           auto tail = waveform.begin();
-           unsigned countdown(window_ - 1);
-           unsigned nCount(0);
-           deque_.clear();
-	  
-           for (; it != waveform.end(); ++it)
-           {
-                while (!deque_.empty() && *it > deque_.back()) deque_.pop_back();
-                deque_.push_back(*it);
-
-                if (countdown > 0) countdown--;
-                else
-                {
-                   if (deque_.front()> minAmp_ && deque_.front()== *std::prev(it,windowPeak_) && *std::prev(it,windowPeak_) != *std::prev(it,windowPeak_-1))
-                   {
-                       int index = int(t0/digiSampling_) + nCount - winOffsetT0_;
-		      
-                       hitList_[index].push_back(FastHit(crId,index,deque_.front()));
-                       if (deque_.front()> minSeedAmp_) seeds_.emplace_back(&(hitList_[index].back()));
-                   }
-
-                   if (*tail == deque_.front()) deque_.pop_front();
-                   ++tail;
-                }
-                ++nCount;
-           } 
+  //----------------------------------------------------------------------------------------------------------
+  void CaloClusterFast::makeClusters(CaloClusterCollection& caloClusters, const art::Handle<CaloHitCollection>& caloHitsHandle)
+  {
+      const Calorimeter& cal = *(GeomHandle<Calorimeter>());
+      const CaloHitCollection& caloHits(*caloHitsHandle);
+      if (caloHits.empty()) return;
+
+      std::vector<int> hits(caloHits.size());
+      std::iota(hits.begin(),hits.end(),0);
+      
+      auto functorTime = [&caloHits](int a, int b) {return caloHits[a].time() < caloHits[b].time();};
+      std::sort(hits.begin(),hits.end(),functorTime);
+      for (auto val : hits) if (caloHits[val].energyDep() < EnoiseCut_) val = -1;
+
+      auto iterSeed = hits.begin();
+      while (iterSeed != hits.end())
+      {
+          //find the first hit above the energy threshold, and the last hit within the required time window
+          const CaloHit& hitSeed = caloHits[*iterSeed];
+          if (*iterSeed==-1 || hitSeed.energyDep()< EminSeed_) {++iterSeed; continue;}
+          double timeStart = hitSeed.time();
+
+          //find the range around the seed time to search for other hits to form clusters
+          auto iterStart(iterSeed), iterStop(iterSeed);
+          while (iterStop  != hits.end()   && (*iterStop==-1  || caloHits[*iterStop].time() - timeStart < deltaTime_))  ++iterStop;
+          while (iterStart != hits.begin() && (*iterStart==-1 || timeStart - caloHits[*iterStart].time() < deltaTime_)) --iterStart;
+          ++iterStart; 
+
+          //start the clustering algorithm for the hits between iStart and iStop
+          std::queue<int> crystalToVisit;
+          std::vector<bool> isVisited(cal.nCrystal()); 
```

Comment:
I don't understand why the isVisited array needs to be so large .  Couldn't you make a struct (int, bool) to combine this with the hits index?  That would both reduce the size and keep the related data close in memory.

Reply:
As I say below, you need to keep track independently of visited crystal ID during a clustering iteration, and hits being used (see below). Clustering typically visits 50-100 crystal IDs with the double ring algorithm (but many IDs are seen several times, so if you include them all regardless of the uniqueness, you have several hundred entries), and you need a "not so small structure" to keep track of them - unless you use a set which is slow. On the other hand, a vector of bool is packed (1 bit / value instead of 8 bits), so you actually need 1800 bits to keep track of all crystals, which isn't that different from O(100) ints. I did some tests and this is the fastest solution I found so far. You might gain O(50 us) is you really optimize like crazy, but I prefer to keep the code simple here.